### PR TITLE
Some InterpreterSelectQuery refactoring

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Non-significant (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Refactoring

### CODE REVIEW

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -635,6 +635,72 @@ static bool allowEarlyConstantFolding(const ExpressionActions & actions, const C
...
+                actions->add(ExpressionAction::removeColumn(column));
+
+            prewhere_info->remove_columns_actions = std::move(actions);
+        }
+
+        columns_to_remove_after_prewhere = std::move(columns_to_remove);
+    }
+    else if (hasFilter())
+    {
+        /// Can't have prewhere and filter set simultaneously
```
Comment (qoega) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp]: Probably add explicit check inside hasPrewhere() block, that hasFilter is not set?

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -635,6 +635,72 @@ static bool allowEarlyConstantFolding(const ExpressionActions & actions, const C
...
+void InterpreterSelectQuery::AnalysisResult::checkActions()
+{
+    /// Check that PREWHERE doesn't contain unusual actions. Unusual actions are that can change number of rows.
+    if (hasPrewhere())
+    {
+        auto check_actions = [](const ExpressionActionsPtr & actions)
+        {
+            if (actions)
+                for (const auto & action : actions->getActions())
+                    if (action.type == ExpressionAction::Type::JOIN || action.type == ExpressionAction::Type::ARRAY_JOIN)
```
Comment (qoega) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp]: Isn't it better to define canChangeRowsNumber helper in ExpressionActions?

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp (hunk 1 above)
Comment (4ertus2) [reply dbms/src/Interpreters/InterpreterSelectQuery.cpp]: It's a refactoring not a rewriting. Do not change any code logic.

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp (hunk 2 above)
Comment (4ertus2) [reply dbms/src/Interpreters/InterpreterSelectQuery.cpp]: It's a refactoring not a rewriting. Do not change any code logic.

**Comment by 4ertus2** [comment]
Local tests with tsan are OK.

**Pull request merged**